[PodLevelResources] Pod Level Hugepage Resources

KevinTMtz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR implements Pod Level Hugepage Resources that require following changes:
1. Add hugepages to pod level supported resources
1. Default pod level hugepage limits
1. Containers with hugepage volume mounts with unset hugepage limits
1. Use pod level hugepage limits for cgroup when unset in container
1. Unit tests for pod level hugepage resources
1. End to end tests for pod level hugepage resources

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
- Changed the Pod API to support `hugepage resources` at `spec` level for pod-level resources.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [Other doc]: https://docs.google.com/document/d/1JaqE2eRmFAPlRayv8vsAWE4SmQCVXQLr9rFPhEaPlvQ/edit?usp=sharing
```

KevinTMtz committed:
Hugepages to pod level supported resources

KevinTMtz committed:
Default pod level hugepage limits

KevinTMtz committed:
Containers with hugepage volume mounts with unset hugepage limits

KevinTMtz committed:
Use pod level hugepage limits for cgroup when unset in container

KevinTMtz committed:
Unit tests for pod level hugepage resources

KevinTMtz committed:
End to end tests for pod level hugepage resources

ndixita commented:
/assign @ndixita

haosdent commented:
Do we need a KEP for this?

ndixita commented:
/test

ndixita commented:
/test all

KevinTMtz committed:
Revert "Pod level hugepage for cgroup when unset in container"

This reverts commit f0a14732371cb6f9521567b3fad564d8a8360bd7 and kuberuntime_container_linux_test.go change.

KevinTMtz commented:
/test all

KevinTMtz commented:
/test all

KevinTMtz committed:
Move pod level hugepage test from e2e to e2e node

KevinTMtz commented:
/test all

KevinTMtz commented:
/test all

KevinTMtz commented:
/retest

KevinTMtz commented:
/retest

ndixita commented:
/milestone 1.33

tallclair commented:
/assign

ndixita commented:
/assign @thockin

lauralorenz commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -1222,6 +1222,298 @@ func TestPodResourcesDefaults(t *testing.T) {
 				},
 			},
+		}, {
+			name:                     "pod hugepages requests=unset limits=unset, container hugepages requests=set limits=set",
```
```suggestion
			name:                     "pod hugepages requests=unset limits=unset, container hugepages requests=unset limits=set",
```

right?

KevinTMtz replied:
Yes, thank you.

lauralorenz commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -1222,6 +1222,298 @@ func TestPodResourcesDefaults(t *testing.T) {
+				},
+			},
+		}, {
+			name:                     "pod hugepages requests=unset limits=set, container hugepages requests=set limits=set",
```
```suggestion
			name:                     "pod hugepages requests=unset limits=set, container hugepages requests=unset limits=set",
```
right?

KevinTMtz replied:
Yes, thank you.

lauralorenz commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -1222,6 +1222,298 @@ func TestPodResourcesDefaults(t *testing.T) {
+				},
+			},
+		}, {
+			name:                     "pod hugepages requests=set limits=set, container hugepages requests=set limits=set",
```
```suggestion
			name:                     "pod hugepages requests=set limits=set, container hugepages requests=unset limits=set",
```
right?

KevinTMtz replied:
Yes, thank you.

lauralorenz commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -23686,62 +23732,81 @@ func TestValidateResourceRequirements(t *testing.T) {
-				core.ResourceName(core.ResourceEphemeralStorage + "2Mi"): resource.MustParse("2Mi"),
+			validateFn: ValidateContainerResourceRequirements,
+		}, {
+			name: "pod resource hugepages without cpu or memory",
```
Is this actually valid? Or is the validation test for container defaults being hoisted up to validate it later?

> When specifying hugepages, it is required to also specify cpu or memory (this will be the same behavior for pod level hugepages)
(from [here](https://docs.google.com/document/d/1JaqE2eRmFAPlRayv8vsAWE4SmQCVXQLr9rFPhEaPlvQ/edit?tab=t.0))

lauralorenz commented on test/e2e_node/hugepages_test.go:
```diff
@@ -481,7 +505,448 @@ var _ = SIGDescribe("HugePages", framework.WithSerial(), feature.HugePages, func
+		ginkgo.JustBeforeEach(func(ctx context.Context) {
+			e2eskipper.SkipUnlessFeatureGateEnabled(features.PodLevelResources)
+
+			// Activate PodLevelResources feature before every test
```
remove comment here

KevinTMtz replied:
Done, thank you.

lauralorenz reviewed (commented):
A few comments from a relative outsider, also FYI to future reviewers I took a close look especially at defaults_test.go against [the docs here](https://docs.google.com/document/d/1JaqE2eRmFAPlRayv8vsAWE4SmQCVXQLr9rFPhEaPlvQ/edit?tab=t.0) and where not otherwise commented they looked right to me 👍

KevinTMtz commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -199,6 +200,7 @@ func SetDefaults_Pod(obj *v1.Pod) {
 	// Pod Requests default values must be applied after container-level default values
 	// have been populated.
 	if utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) {
+		defaultPodLimits(obj)
```
Done, I agree, thank you.

KevinTMtz commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -405,10 +406,19 @@ func getPageSizeMountOption(medium v1.StorageMedium, pod *v1.Pod) (string, error
+	}
+
+	for _, resourceList := range podLevelAndContainerLevelRequests {
 		// We can take request because limit and requests must match.
```
Done, thank you.

ndixita commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -456,7 +458,7 @@ func defaultPodRequests(obj *v1.Pod) {
 	// default to the effective requests of all the containers for that resource.
 	for key, aggrCtrLim := range aggrCtrReqs {
-		if _, exists := podReqs[key]; !exists && resourcehelper.IsSupportedPodLevelResource(key) {
+		if _, exists := podReqs[key]; !exists && resourcehelper.IsSupportedPodLevelResource(key) && !corev1helper.IsHugePageResourceName(key) {
```
Please add a comment before the if statement stating why for huge pages the requests are not set equal to aggregated container requests

thockin reviewed (commented):
API related bits seem OK.

HugePages has always felt clunky and this continues the tradition.

thockin commented:
Ping me for final approval when kubelet reviews are done.

tallclair commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -56,14 +58,14 @@ type PodResourcesOptions struct {
 
 func SupportedPodLevelResources() sets.Set[v1.ResourceName] {
-	return supportedPodLevelResources
+	return sets.New(supportedPodLevelResources.UnsortedList()...).Insert(v1.ResourceHugePagesPrefix)
```
nit: prefer

```suggestion
	return supportedPodLevelResources.Clone().Insert(v1.ResourceHugePagesPrefix)
```

KevinTMtz replied:
Done, thank you.

tallclair reviewed (commented):
Does https://github.com/kubernetes/kubernetes/blob/648620da0ebdc87af8053d564d648c12c6606231/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go#L146 need to be updated to fallback to pod-level resources?

KevinTMtz commented:
> Does
> 
> https://github.com/kubernetes/kubernetes/blob/648620da0ebdc87af8053d564d648c12c6606231/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go#L146
> 
> need to be updated to fallback to pod-level resources?

Currently we are not going to set the cgroup values for the containers that do not specify hugepage limits, because of the starving problem caused by not having hugepages isolation between the containers that specify limits and the ones that do not, when setting the HugeTLB cgroups to max (or max minus the aggregated limits) of the containers that did not specify.
If we do not set the hugepage cgroup to the container with unset request/limits, the behavior will depend of the containerd HugeTLB being enabled or not:

- Containerd HugeTLB Enabled
    - The containers with hugepage request/limit will not be starved, their limits will be respected (because only them will be able to use the hugepages, and the available hugepages will be the aggregated sum of their request/limit)
    - The containers without hugepage request/limit will not have access to any hugepages

- Containerd HugeTLB Disabled
    - The containers with hugepage request/limit set and unset will have the same behavior, all of them will be able to access the whole hugepage resources in the pod, all of them will be able to starve all the others

Initially, I added the commit [Use pod level hugepage limits for cgroup when unset in container](https://github.com/kubernetes/kubernetes/pull/130577/commits/6138dadff59f4c6c8920d2ab673aaa27baa1f69a), however because of this situation, I reverted the change

@tallclair

ndixita commented:
@KevinTMtz Is it allowed for aggregated huge pages limits to be greater than pod-level huge pages limits? Did we test this case?

KevinTMtz commented:
> @KevinTMtz Is it allowed for aggregated huge pages limits to be greater than pod-level huge pages limits? Did we test this case?

It is not possible, since hugepages can not be overcommitted.

If the user tries to do something like this:
```
apiVersion: v1
kind: Pod
metadata:
  name: pod-level-hugepages
spec:
  resources:
    limits:
      memory: 256Mi
      hugepages-2Mi: 512Mi
  containers:
  - name: hptest-1
    image: "hptest:0.1"
    resources:
      limits:
        cpu: 1
        hugepages-2Mi: 514Mi
    volumeMounts:
    - mountPath: /hugepages-2Mi
      name: hugepage-2mi
  volumes:
  - name: hugepage-2mi
    emptyDir:
      medium: HugePages-2Mi
```

This error will be obtained:
```
The Pod "pod-more-container" is invalid: 
* spec.resources.requests[hugepages-2Mi]: Invalid value: "512Mi": must be greater than or equal to aggregate container requests of 514Mi
* spec.resources.containers[0][hugepages-2Mi].limits: Invalid value: "514Mi": must be less than or equal to pod limits of 512Mi
```

@ndixita

tallclair commented:
/lgtm
/approve

thockin commented:
/approve

liggitt commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -56,14 +58,14 @@ type PodResourcesOptions struct {
 // the resource is supported.
 func IsSupportedPodLevelResource(name v1.ResourceName) bool {
-	return supportedPodLevelResources.Has(name)
+	return supportedPodLevelResources.Has(name) || strings.HasPrefix(string(name), v1.ResourceHugePagesPrefix)
```
a couple comments here:
1. this is effectively relaxing validation, which means we have to make sure it isn't active by default in the first release it is present in (1.33 will choke on pods this allows in with pod-level hugepages requests/limits)
2. this allowed things into the API which weren't actually implemented or tested yet (implementation is in https://github.com/kubernetes/kubernetes/pull/131089) ... we generally want to make sure implementation merges with API

both of those issues need to be addressed before pod-level resources enables by default

liggitt replied:
oh, I just realized this merged in 1.33, not 1.34. whew

tallclair replied:
Kevin also clarified that this PR did implement support. The implementation is implicit - pod level huge pages were already set on the pod cgroup using the Pod{Requests,Limits} helper which already factors in pod-level resources, so once we allow huge-pages to be set at the pod-level, the implementation is already there.
